e2e_test.ts

```diff
@@ -624,3 +631,101 @@ Deno.test("[e2e] GET /notifications/[id]", async (test) => {
     assertResponseNotFound(resp);
   });
 });
+
+Deno.test("[e2e] POST /api/comments", async (test) => {
+  const url = "http://localhost/api/comments";
+  const item = genNewItem();
+  await createItem(item);
+
+  const formData = new FormData();
+  formData.append("text", "Comment text");
+  formData.append("item_id", item.id);
+
+  await test.step("returns HTTP 401 Unauthorized response if the session user is not signed in", async () => {
+    const resp = await handler(
+      new Request(url, { method: "POST" }),
+    );
+    assertFalse(resp.ok);
+    assertEquals(resp.status, Status.Unauthorized);
+  });
+
+  const user = genNewUser();
+  await createUser(user);
+
+  await test.step("returns HTTP 400 Bad Request response if comment is missing text", async () => {
+    const invalidFormData = new FormData();
+    invalidFormData.append("item_id", item.id);
+    const resp = await handler(
+      new Request(url, {
+        method: "POST",
+        headers: { cookie: "site-session=" + user.sessionId },
+        body: invalidFormData,
+      }),
+    );
+
+    assertEquals(resp.status, Status.BadRequest);
+  });
+
+  await test.step("returns HTTP 400 Bad Request response if comment is missing item_id", async () => {
+    const invalidFormData = new FormData();
+    invalidFormData.append("text", "Comment text");
+    const resp = await handler(
+      new Request(url, {
+        method: "POST",
+        headers: { cookie: "site-session=" + user.sessionId },
+        body: invalidFormData,
+      }),
+    );
+
+    assertEquals(resp.status, Status.BadRequest);
+  });
+
+  await test.step("returns HTTP 404 Not Found response if the item is not found", async () => {
+    formData.set("item_id", "not-found-item-id");
+    const resp = await handler(
+      new Request(url, {
+        method: "POST",
+        headers: { cookie: "site-session=" + user.sessionId },
+        body: formData,
+      }),
+    );
+
+    assertEquals(resp.status, Status.NotFound);
+  });
+
+  await test.step("a comment is created by the same session user, a new notification is not created", async () => {
+    formData.set("item_id", item.id);
+    const createCommentSpy = spy(createComment);
+    const createNotificationSpy = spy(createNotification);
+    const resp = await handler(
+      new Request(url, {
+        method: "POST",
+        headers: { cookie: "site-session=" + user.sessionId },
+        body: formData,
+      }),
+    );
+
+    assertEquals(resp.status, Status.SeeOther);
+    assertSpyCallAsync(createCommentSpy, 0);
+    assertSpyCalls(createNotificationSpy, 0);
```

Comment:
Calling this out, this is a way to detect that a spy is not called by testing the number of calls that were made to it.. whereas `assertSpyCall` and `assertSpyCallAsync` is actually testing that an `index` call is being made. 

Since a new `ulid()` is created when a comment is built, I wasn't able to find a way around checking the `args` provided to the spy, just that it was called.

Reply:
I used `assertSpyCalls()` in all instances for `createComment()` and `createNotification()`. Frankly, the code looks correct to me. However, tests are failing, saying that both test steps never call `createComment()` and `createNotification()`, which seems incorrect. Am I missing something here? @kt3k, does this look right to you?

Reply:
Odd.. yea, I tested a bit more to validate my misunderstanding here.. for instance;

`createCommentSpy.calls` has nothing..even though `assertSpyCallAsync` seems like it is "working", and now I couldn't be sure... Dang, I'll dig in a little bit to see what's wrong here.

Reply:
I think part of the problem here is that `createComment` (for instance) is also imported into `e2e_test.ts` to facilitate test setup. I tried to spy on `kv.atomic` and that spy wasn't working in the context of this test either. Hrm!

Reply:
Rethinking the test and using a spy here.. I added a commit that replaces the spy on these two tests with queries and existence checks. Let me know what you think?